ENH: Add dtype argument to random.randint.

charris opened the pull request:
~~This is a WIP, in particular, the examples need fixing, tests need to be added, and a decision should be made whether to access these functions through a dtype argument in the current `random_integers` function. If the latter is chosen rk_functions for uint8, uint16, and maybe bool need to be added for efficiency, on the other hand, less new documentation needs to be written ;)~~

The following dtypes are now available from the new `dtype` argument added to randint.

```
* np.bool,
* np.int8, np.uint8,
* np.int16, np.uint16,
* np.int32, np.uint32,
* np.int64, np.uint64,
* np.int_ (long), np.intp
```

Tests for all the new functionality are needed, but things seem to work. It should also be almost trivial to template the code added to mtrant.pyx, the 9 added functions only differ in using different types.

When merged #6812 will be closed and the integer part of #6790 will be done..

~~Alternate PR's addressing this are #6824 and #6869, but I like this one as far as it goes.~~

charris committed:
ENH: Add dtype argument to random.randint.

Random ndarrays of the following types can now be generated:

* np.bool,
* np.int8, np.uint8,
* np.int16, np.uint16,
* np.int32, np.uint32,
* np.int64, np.uint64,
* np.int_ (long), np.intp

The specification is by precision rather than by C type. Hence, on some
platforms np.int64 may be a `long` instead of `long long` even if the
specified dtype is `long long` because the two may have the same
precision. The resulting type depends on which c type numpy uses for the
given precision. The byteorder specification is also ignored, the
generated arrays are always in native byte order.

The dtype of the result could be made more explicit if desired without
changing the user visible results.

charris committed:
TST,BUG: Fix use of randint in test_mem_overlap.py.

The default randint function returns a C long type which does not have
enough range to test indexes on Windows 64. The fix here is to use
specify a np.intp dtype for the randint call now that we have that
option.

Closes #6812.

gfyoung commented:
There seems to be a lot of boilerplate.  Is it possible to condense?  Especially with other issues like broadcasting (see #6902) or ```dtype``` (see #6869), it seems like this could easily blow up with so many functions implementing <i>nearly</i> the same thing but with 32-bit or 64-bit ```int```.

Also, is there a way to test / demonstrate the issue with the random stream that is apparently rampant in the current implementation of ```random_integers```?  That would be quite useful (at least for my sake, given that I have been looking at this for so long) to illustrate why these changes are necessary.

rkern commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1426,15 +1435,271 @@ cdef class RandomState:
+        """
+        random_int32(low, high=None, size=None)
+
+        Return random npy_intp integers between `low` and `high`, inclusive.
```
`int32`?

charris commented:
@gfyoung You can implement both the dtype and broadcasting using the added functions. You need both 32 and 64 to implement both intp and int with the same functions, the same functions added to randomkit (which uses unsigned) can also be used for unsigned. To do a efficient job with smaller integer types more functions need to be added to randomkit for efficient implementation, i.e., there are two int16 in one int32 sample.

Not clear on what you consider boiler plate besides the documentation...

gfyoung commented:
@charris : Regarding ```dtype``` and broadcasting, that actually wasn't my question.  I'm not worried about ability to implement.  No, no, I know both are certainly possible, broadcasting to say the least given how many times it has been done already in ```mtrand.pyx```.

My concern was that won't you have to implement everything at least twice now in your versions?  Maybe I'm getting my semantics / definitions wrong, but your two functions are boilerplate given how similar they are to each other.

Also, you say "you need" to implement the functions separately for 32-bit and 64-bit.  Again, is there a way to test or illustrate the shortcomings?  Also, why is it that you are using the 32-bit Mersenne Twister for ```npy_uint64```?  From what I remember, @rkern said on my PR that this should not be done, and that there is a completely separate algorithm for 64-bit.

I unfortunately can't pull down the branch right now to try it myself, but how does the change address the failed method call in #6812 with ```np.iinfo(np.int64).max```?  Do you have to call ```random_int64```, or does ```random_intp``` suffice?

rkern commented:
That's not what I said. What I said was that just changing `unsigned long` to `unsigned long long` in the core PRNG would not help you. Instead, I suggested exactly what @charris implemented here, to create `rk_uint64` that generates a `uint64` from two draws from the core PRNG.

charris commented:
@gfyoung You need 64 bit random numbers in order to generate 64 bit random numbers. OTOH, it is inefficient to use 64 bit numbers to generate 32 random numbers. The two functions are similar, but not the same. Note that `rk_random_uint64` will use 32 or 64 bit random samples depending on the size of the range, and we need to preserve that in order to retain backwards compatibility.

It would help if you would make specific comments about specific lines/functions. It it difficult to respond to generalities.

gfyoung commented:
@rkern: Oh, sorry!  You're right.  Memory failed me there.  By the way in the original code (i.e ```master``` branch), doesn't ```rk_ulong``` do exactly that, which is called from ```rk_interval```, which in turn is called by ```random_integers```?  If not, then how was it then that my added test in #6824 passed both on my machine and on 64-bit Linux on Travis?

@charris: Apologies that I haven't been able to give more specific pointers.  While I realize this is WIP, my comments have been directed at the overall organization and structure of the code.  My concern was directed towards the future additions of ```dtype``` and broadcasting, which would have to be added in both functions.  I'm just trying to understand why you organized the code the way you did, if that's okay.

Also, if you could address the question regarding the ```np.iinfo(np.int64).max``` call, that would be great as well!

charris commented:
@gfyoung If we go the dtype route, it only needs to be added to one function, which will call the appropriate others, which may be hidden (or not). But we do need the others, as they will all differ in lower level details.

I don't see an `np.iinfo(np.int64).max` call here, what are you referring to?

gfyoung commented on numpy/core/tests/test_mem_overlap.py:
```diff
@@ -98,6 +98,7 @@ def test_overlapping_assignments():
 def test_diophantine_fuzz():
     # Fuzz test the diophantine solver
     rng = np.random.RandomState(1234)
+    random_intp = rng.random_intp
```
Nitpick: ```random_intp``` is on the closed interval, but ```randint``` is on the half-open interval.  Perhaps this is not so consequential since the tests still pass, but you are changing the function being called.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1426,15 +1435,272 @@ cdef class RandomState:
+            msg = 'Unimplemented for sizeof(long) == %d' % sizeof(long)
+            raise RuntimeError(msg)
+
+    def random_intp(self, low, high=None, size=None):
```
Can this just be ```random_integers``` instead?

charris replied:
No. One is C `long`, the other is `Py_ssize_t` and the two are not always the same, that is why there is an error on windows 64 where C `long` is 32 bits and `Py_ssize_t` is 64 bits.

gfyoung replied:
But this function prevents the overflow error, and on remaining platforms (and if there are more, please correct me on this!), ```long``` is the same size of ```Py_ssize_t```.  Given that the difference is significant on few platforms, I would think that the difference between the two is minimal and therefore confusing for users.

charris replied:
Not really. They have different uses, `np.intp` is the indexing integer, and we need to keep `np.int` as is for backwards compatibility. Python 2 had two integer types, `int`, and `long`, not the same as the C types of the same name. The `int` type was implemented as C `long`, which was efficient, and the later `long` type has arbitrary precision. Python 3 has only the `long` type, but we need to carry along the historic crud so as to not break old code/results.

gfyoung replied:
Okay, that makes sense.  Could that be noted somewhere?  That would be good to know.

charris replied:
You have a point, in that the way things are implemented the returned values would not change for the same ranges if only `np.intp` were used, but the array type would, doubling in memory footprint on windows, and the returned scalar values would be of `long` python type for python 2.7 on 64 bit Windows.

gfyoung replied:
That is true.  I wonder then if you could put logic in ```random_integers``` to check the OS and Python version and then call the appropriate one (either the ```np.intp``` or ```long``` versions) to maintain backwards compatibility?  Then the user wouldn't have to worry / think about which one to call.  This would require abstracting the current code in ```random_integers``` into say a function called ```rand_int``` (placeholder, this name may not work so well as it would get confusing with ```randint```; though perhaps the names might need to be refactored).

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1426,15 +1435,272 @@ cdef class RandomState:
+        elif sizeof(npy_intp) == sizeof(npy_int32):
+            return self.random_int32(low, high, size)
+        else:
+            msg = 'Unimplemented for sizeof(npy_intp) == %d' % sizeof(long)
```
```np_intp``` instead of ```long```?

charris replied:
Yes.

gfyoung commented:
@charris : Ah, I see.  Just add it to ```random_integers``` and then branch out to the appropriate calls in ```randomkit.c```.  For efficiency purposes, that's fair.

The ```np.iinfo(np.int64).max``` call I'm referring to is in the first comment of @jreback issue #6812, which technically is not the failing ```mem_overlap``` test failure as you described in your PR description.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1426,15 +1435,272 @@ cdef class RandomState:
+        else:
+            lo = low
+            hi = high
+
```
I wonder if we could wrap this ```if...else``` block into a ```try...except``` block to provide a useful message about ```OverflowError``` issues that we are seeing in #6812, something like:
```
try:
... # if-else block
except OverflowError:
raise ValueError("Boundary values too large for int32.  Please use random_int64 instead.")
```

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1426,15 +1435,272 @@ cdef class RandomState:
+        if sizeof(long) == sizeof(npy_int64):
+            return self.random_int64(low, high, size)
+        elif sizeof(long) == sizeof(npy_int32):
+            return self.random_int32(low, high, size)
```
I'm not sure if this will break backwards compatibility, but would logic checking the boundaries inputs be useful?  For example, they are in the range [```np.iinfo(np.int32).min, ```np.iinfo(np.int32).max```], then call ```random_int32```?  Ditto for ```random_intp```.

sturlamolden commented on numpy/random/mtrand/randomkit.c:
```diff
@@ -240,6 +246,101 @@ rk_random(rk_state *state)
+rk_uint64(rk_state *state)
+{
+    npy_uint64 upper = (npy_uint64)rk_random(state) << 32;
+    npy_uint64 lower = (npy_uint64)rk_random(state);
```
```rk_random```return unsigned long, so should we mask out the upper bits?
```C
npy_uint64 lower = (npy_uint64)rk_random(state) & 0xFFFFFFFF;
```

charris replied:
Upper bits are zero, `rk_random` is always a 32 bit number stored in a long. Masking won't hurt, but shouldn't be needed.

sturlamolden replied:
If we know they are zero it is ok. I was worried they might have junk bits which could interfere with the distribution and the determinism.

charris replied:
I understand ;) The Mersenne twister would be in big trouble if it weren't true, as the output would depend on the size of long. It is hard to trace all the way through, but one must have faith. Sometimes. In any case, the current `rk_ulong` function makes the same assumption, although it uses a `|` instead of a `+`. Hmm, might be a trifle faster.

sturlamolden commented on numpy/random/mtrand/randomkit.c:
```diff
@@ -240,6 +246,101 @@ rk_random(rk_state *state)
+NPY_INLINE static npy_uint32
+rk_uint32(rk_state *state)
+{
+    return (npy_uint32)rk_random(state);
```
```C
return (npy_uint32)(rk_random(state) & 0xffffffffUL);
```

charris replied:
Not needed, the cast is defined to discard the higher bits by the C standard.

rkern commented:
Generally speaking, the `high`-inclusive `random_integers(low, high)` is considered [an API mistake](https://mail.python.org/pipermail/python-ideas/2015-September/036434.html). It's only here because it was in Numeric's `RandomArray`. I'd prefer to expand only `randint()` to other dtypes instead.

gfyoung commented:
@rkern : That's interesting that you point that out.  What happens if people wanted to generate numbers that included the maximum possible integer (e.g. ```np.iinfo(np.int64).max```)?

rkern commented:
That just means that we have to be careful about dealing with the `high` argument that is input, but it should be manageable.

gfyoung commented:
@rkern : Oh, that was actually more of a theoretical question than a question directly related to the changes here.  In any case, I'm not sure that I follow your comment regarding only implementing ```dtype``` for just ```randint``` and not ```random_integers``` since ```randint``` calls ```random_integers``` currently.

As an aside, perhaps once all of the refactoring has been done, it could be worthwhile to deprecate ```randint``` and ```random_integers``` with just one function that accepts a ```is_half_open``` defaulted to ```True```.  Just throwing that out there.

rkern commented:
It's the other way around. `random_integers()` calls `randint()`.

gfyoung commented:
@rkern : They have been reversed (#6885).  Also, I had a question for you about the two draws from the PRNG, which I think got lost in the discussion, so I'll paste it here:

By the way in the original code (i.e master branch), doesn't rk_ulong do exactly that, which is called from rk_interval, which in turn is called by random_integers? If not, then how was it then that my added test in #6824 passed both on my machine (32-bit) and on 64-bit Linux on Travis?

charris commented:
@rkern OK, things can be fixed up. To be clear, `randint` is the preferred function. ISTM that in that case we should deprecate `random_integers` or at least document that `randint` should be used instead. Perhaps the proposed `dtype` argument should only be added to `randint` as well to encourage that.

rkern commented:
@gfyoung `rk_ulong()` *does* make two draws when `unsigned long` is 64 bits wide. That's what that `#if ULONG_MAX <= 0xffffffffUL` test is controlling.

njsmith commented:
Regarding the nitty-gritty details of the various method semantics: I do think it would indeed be nice if there were some way to sample a uniform `uint64`, though it's a bit tricky to get the API right because it's a weird corner case that most users probably won't care about. I also agree that the current `random_integers`/`randint` situation is confusing. An additional extra-confusing aspect that hasn't been mentioned yet in this thread is that the stdlib random module also has [functions for sampling from both half-open and closed intervals](https://docs.python.org/3/library/random.html#random.randrange), and it also uses the name `randint` for one of those functions, but they do it backwards from us -- their `randint` is our `random_integers`, and our `randint` is their `randrange`.

Proposal: soft deprecate *both* `randint` and `random_integers` (by "soft deprecate" I mean: document that you shouldn't use them, mayyyybe emit a warning when they're used, but don't make any plans to actually remove them), and replace them both with a new method `randrange([start,] stop, [step, closed=False, dtype=int])`. Rationale: the `randrange` name is more transparent to users who already know `range` and more consistent with the stdlib; the new `closed=` kwarg (as suggested by @gfyoung) provides an elegant way to sample from the full fixed-width integer range without cluttering up the top-level documentation and forcing users to understand the subtleties of how the half-open convention interacts with fixed-width integers just to get started.

If that sounds good to y'all then someone should post it to the list I guess...

Two things that might benefit from further refinement: (a) the `closed=` name is somewhat jargony for parts of our user base, maybe there's something even better? (`include_max=`?), (b) do we want to take advantage of the addition of this new method to clean up the default integer type?

rkern commented:
I'd suggest fixing the technical issues with the existing function of interest first and then do the renaming for clarity separately. 

As for the `rangerange()` proposal itself, I'd actually omit `step=` as a foolish consistency with `random.randrange()` which is itself a foolishly consistent with `range()`. This should be a fairly primitive function, so making it do more things with more options is a pessimization in terms of usability, IMO. As for `closed=`, there isn't a problem with specifying the full range with the open-interval convention; the caller is not restricted to passing in fixed-width integers.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -885,15 +1013,15 @@ cdef class RandomState:
         """
-        randint(low, high=None, size=None)
+        randint(low, high=None, size=None, dtype=np.int)
 
```
Documentation mismatch on the ```dtype``` param?

gfyoung replied:
Documentation mismatch with ```dtype``` param?

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -936,14 +1069,24 @@ cdef class RandomState:
+
+        with self.lock:
+            return randfunc(low, high - 1, size, self.state_address)
 
```
Ah, this is how you got around the issue that I was addressing in #6885.  Cool.

gfyoung commented:
Just curious, when we say the "full" range, what range of numbers are we referring to?

@rkern : I agree with saving the renaming until later.  I was just suggesting that as a potential next step because the names ```randint``` and ```random_integers``` really don't explain how they're different from each other.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -936,14 +1069,24 @@ cdef class RandomState:
+        if high > highbnd:
+            raise ValueError("high is out of bounds for %s" % key)
+        if low >= high:
+            raise ValueError("low >= high")
```
This is not within the scope of this PR, but I was thinking that ```randint``` could one day implement something similar to ```numpy.array``` in which it tries to determine the minimum ```dtype``` required to generate numbers in the range requested.  Throwing that out as another idea for future "expansion" FLOABT.

charris committed:
BUG: Get rid of C++ style comment in multiarray_tests.c.src

charris commented:
The failing debug build looks like a cython bug. The warning causing the failure is
```
numpy/random/mtrand/mtrand.c:29108:3: warning: this decimal constant is unsigned only in ISO C90 [enabled by default]
```
and AFAICT, it is the cython generated code
```
__pyx_int_neg_2147483648 = PyInt_FromLong(-2147483648L)
```
A relevant stackoverflow discussion at http://tinyurl.com/np5g5de. Apparently the constant is `2147483648L`, which is out of bounds for long on i386, and the minus makes it an expression. That is `- 2147483648L` is trying to negate a `long` that doesn't exist. Heh. How unfortunate. I don't see an easy way to work around that. Changing the C standard makes no difference, which in retrospect makes sense.

charris commented:
Hmm, c99 does seem to fix the compilation, but test errors result. Ignoring the warning results in the tests passing.

I'd be curious for comments about the design here for the added `dtype` keyword.

gfyoung commented on numpy/random/mtrand/randomkit.c:
```diff
@@ -240,6 +246,101 @@ rk_random(rk_state *state)
+
+
+/*
+ * Fills an array with cnt random npy_uint64 between off and off + rng
```
I think you meant ```npy_uint32``` here?

charris replied:
For some reason that fix doesn't show up here.

gfyoung replied:
Ah, yes, I see.  You did indeed make that change.  Congratulations on breaking GitHub :smile:

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -574,6 +579,128 @@ def _shape_from_size(size, d):
+        out = <npy_uint32 *>PyArray_DATA(array)
+        with nogil:
+            rk_random_uint32(off, rng, cnt, out, state)
+        return array
```
Question: why is everything in the ```else``` block (save ```array```) being casted as ```np.uint32``` when we are returning a random array of ```np.int32```?  Similar question for ```rand_int64```.

jaimefrio replied:
Because it works, the magic of two's complement I guess... Say you pass in a range of -5 to 5, and for brevity let's pretend this was an 8 bit version. The range is properly converted to a uint8 10, but low gets converted to uint8 251, 11111011 in binary. When the `rk_random_uint32` function e.g. adds 6, 00000110 in binary, to low, the result would be 257, 100000001 in binary, but because we have fixed width integers, and unsigned addition overflow is not undefined, the high bit is discarded, and you get the correct 1 value.

Perhaps it is worth a comment somewhere: it wasn't obvious to me either.

gfyoung replied:
Ah, okay, that's fair.  I think @charris made comments about numbers wrapping in ```randomkit.c``` in the case that ```off``` is large.  Though it would seem useful to make a minor comment somewhere in ```mtrand.pyx``` too as you suggested.

charris replied:
At the bit level, there is no difference between signed/unsigned addition and subtraction on two's complement machines. However, unsigned arithmetic is specified to be modular, so overflow is always ignored. For signed integers the behavior may be undefined. Here we need to be careful to convert signed integers to two's complement signed binary integers and then treat those integers as unsigned, hence the double cast in some cases. The difference needs to be a positive so the generated random offsets can the added to the offset. To see how this works in practice, think of the unsigned numbers a located on a circle, increasing clockwise, with 0 just after the largest, and call 0 the wrapping point. Signed arithmetic is on the same circle, but the wrapping point is chosen to be at the most negative value represented by `0x100....0` on the circle. Addition of positive numbers moves a point clockwise around the circle in both cases. To make sure every thing works as expected the low and high bounds are checked before the call so that proceeding clockwise from the lowest to the highest never crosses the chosen wrapping point.

Note that another option for representing signed numbers is offset binary, which can be advantageous in some circuit designs and DSP.

gfyoung replied:
Right, that is true.  What you did is perfectly fine.  I was just confused initially because the return type and castings were slightly different.  Thanks @charris and @jaimefrio for clarifying!

rkern commented:
> Just curious, when we say the "full" range, what range of numbers are we referring to?

@gfyoung We mean that for a given requested `output_dtype` we can request a range anywhere from `np.iinfo(output_dtype).min` through `np.iinfo(output_dtype).max`, inclusive.

jaimefrio commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -574,6 +579,128 @@ def _shape_from_size(size, d):
+        Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
+        ``m * n * k`` samples are drawn.  Default is None, in which case a
+        single value is returned.
+
```
The `state` parameter is missing here, but not in the 64 bit version. Probably worth templating at least the docstring, both to avoid mismatches, and to make life easier to whoever eventually creates the lower bit count versions.

gfyoung replied:
Judging from the email discussion, it seems though people are more in favor to implement the lower bit count versions in ```randomkit.c``` and hence to avoid extra documentation as @charris mentioned I think in his PR description.

Nevertheless, if my comments can attest to anything, I do agree that the documentation does seem kind of shaky right now.  Though it is WIP, and since there is a lot of functionality left to implement (assuming that @charris intends to add ```dtype``` functionality in full), the documentation could probably be tightened up at the end in case anything changes along the way.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -574,6 +579,128 @@ def _shape_from_size(size, d):
+def rand_int64(low, high, size, rngstate):
+    """
+    random_int64(low, high=None, size=None)
+
```
Function signature mismatch

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -574,6 +579,128 @@ def _shape_from_size(size, d):
+def rand_int32(low, high, size, rngstate):
+    """
+    random_int32(self, low, high=None, size=None, state)
+
```
Function signature mismatch

charris committed:
TST: Increase the allowable warning count to 1 for i386 debug test.

Cython generated C code contains the number '-2147483648L', which
leads to a warning on 32 bit platforms:

"Warning: this decimal constant is unsigned only in ISO C90"

See the discussion at http://stackoverflow.com/questions/9941261/

The compiled code seems to run correctly despite the warning and
if there are problems, they should turn up in the nose testing.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -574,6 +586,304 @@ def _shape_from_size(size, d):
+        out = <npy_bool *>PyArray_DATA(array)
+        with nogil:
+            rk_random_uint8(off, rng, cnt, out, state)
+        return array
```
Did you mean to call ```rk_random_bool``` here instead?

charris replied:
Oops, good catch, thanks. More that doubles the speed to use the right function.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1356,11 +1682,13 @@ cdef class RandomState:
         """
```
Just curious: in light of the recent discussion surrounding half-open and closed intervals, is it worth deprecating this function in this release?

charris committed:
TST: Enable AppVeyor failure to be reported.

AppVeyor failures were filtered out until numpy issues were fixed.
The last issues were the test_mem_overlap failures on 64 bit windows,
and those are fixed.

gfyoung commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -908,6 +1219,11 @@ cdef class RandomState:
+            name, i.e., 'int64', 'int`, etc, so byteorder is not available
+            and a specific precision may have different C types depending
+            on the platform.
 
```
Should probably mention that ```dtype``` is optional and that it defaults to `l`.

charris replied:
Done.

charris committed:
TST: Add tests for new randint functionality.

* check exceptions
* check extreme bounds are reachable
* check that all values are in the specified bounds
* check repeatability of sequences

More exact statistical tests would be nice, but that is another
project.

charris committed:
DOC: Document new randint dtype parameter in 1.11.0 release notes.

charris commented:
OK, I think this is ready. Timing generation of 10**6 random longs, is about  11.5 ms for [0,5) range, ~30% faster than before. Same number of booleans is 863 µs. The timing vary depending on the range chosen. Powers of two for the upper (non inclusive) bound are good, powers of two + 1 are bad.
```
In [10]: timeit random.randint(0, 8, size=10**6, dtype=np.uint64)
100 loops, best of 3: 3.82 ms per loop

In [11]: timeit random.randint(0, 9, size=10**6, dtype=np.uint64)
100 loops, best of 3: 13.6 ms per loop
```

Previous timings (64 bit longs)
```
In [1]: timeit random.randint(0, 8, size=10**6)
100 loops, best of 3: 6.46 ms per loop

In [2]: timeit random.randint(0, 9, size=10**6)
100 loops, best of 3: 16.4 ms per loop
```
So a 20%-40% speedup.

gfyoung commented:
@charris : There are still several documentation issues that should probably be fixed (see the comments I made that have been collapsed yet in an outdated diff info).

charris commented:
An indication of just how much a jump backward in a loop costs.
```
In [13]: timeit np.random.randint(0, 128, size=10**6, dtype=np.uint64)
100 loops, best of 3: 3.8 ms per loop

In [14]: timeit np.random.randint(0, 65, size=10**6, dtype=np.uint64)
100 loops, best of 3: 15.5 ms per loop
```
If the jump overhead was zero, then the second timing would be 2x the first, but as there is a jump back before the second draw that has to account for the extra time. That suggests that the speed could be increased by perhaps about a factor of two by unrolling the loop a bit.

charris commented:
OK, I'm going to put this in so that the appveyor failures can show.

rkern commented on doc/release/1.11.0-notes.rst:
```diff
@@ -86,6 +86,23 @@ New Features
+  The specification is by precision rather than by C type. Hence, on some
+  platforms np.int64 may be a `long` instead of `long long` even if the
+  specified dtype is `long long` because the two may have the same
+  precision. The resulting type depends on which c type numpy uses for the
```
`C`

charris replied:
Heh, OK, I'll fix that in the notes revisions coming up.

rkern commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -908,6 +1219,13 @@ cdef class RandomState:
             single value is returned.
+        dtype : dtype, optional
+            Desired dtype of the result. All dtypes are determined by their
+            name, i.e., 'int64', 'int`, etc, so byteorder is not available
```
Quotes.
